modules/internal/ChapelDomain.chpl

```diff
@@ -1283,6 +1283,271 @@ module ChapelDomain {
+                      'manager initialized with \'checks=false\'';
+          compilerError(msg);
+          return false;
+        }
+      }
+
+      pragma "no doc"
+      proc _checkIfAllElementsAreInitialized(baseArr) {
+
+        // Arrays of non-nilable classes.
+        if baseArr.chpl_isElementTypeNonNilableClass() {
+          for idx in newIndices() {
+            if baseArr.chpl_unsafeAssignIsClassElementNil(this, idx) {
+              baseArr.chpl_unsafeAssignHaltUninitializedElement(idx);
+              halt();
+            }
+          }
+
+        // Arrays of non-default-initializable types.
+        } else if !baseArr.chpl_isElementTypeDefaultInitializable() {
+
+          // TODO: For non-default-initializable elements, we'd iterate
+          // through a bitvector mapped to newly added slots.
+          halt('Not implemented yet!');
+
+        // Should never reach here.
+        } else {
+          halt('Internal error!');
+        }
+      }
+
+      pragma "no doc"
+      proc _ensureNoLongerManagingThis() {
+        if !_isActiveManager then return; else _isActiveManager = false;
+
+        // Possible runtime checks, reset the resize policy of owned arrays.
+        for baseArr in _lhsBaseArrays() {
+          if !baseArr.chpl_isElementTypeDefaultInitializable() {
+            if _checks then _checkIfAllElementsAreInitialized(baseArr);
+            const policy = chpl_DdataResizePolicy.Normal;
+            baseArr.chpl_setResizePolicy(policy);
+          }
+        }
+      }
+
+      proc deinit() {
+        _ensureNoLongerManagingThis();
+      }
+
+      pragma "no doc"
+      proc _isArrayOwnedByLhsDomain(arr) {
+        for baseArr in _lhsBaseArrays() do
+          if baseArr == chpl__getActualArray(arr) then return true;
+        return false;
+      }
+
+      pragma "no doc"
+      proc _moveInitializeElement(arr, idx, in value) {
+        import Memory.Initialization.moveInitialize;
+        ref elem = arr[idx];
+        moveInitialize(elem, value);
+      }
+
+      pragma "no doc"
+      proc _checkThatIndexMatchesArrayShape(arr, idx) {
+      }
+
+      /*
+        Initialize a newly added array element at an index with a new value.
+        If the array element at `idx` has already been initialized, this
+        method will silently perform assignment instead. This method will
+        error if `idx` is not a valid indice in `arr`.
+      */
+      proc initialize(arr: [?d], idx, in value: arr.eltType) {
+
+        // Check to make sure value and array element types match.
+        if arr.eltType != value.type then
+          compilerError('Initialization value type \'' + value:string +
+                        '\' does not match array element type \'' +
+                        arr.eltType:string + '\'');
+
+        // Only certain forms of index are permissable.
+        _checkThatIndexMatchesArrayShape(arr, idx);
+
+        // TODO: Is this even worth checking? It's O(n) number of arrays.
```

Comment:
If you modify the implementation it'll be an O(1) check.

Reply:
I didn't think the O(n) part of this was `chpl__getaActualArray()`, I thought the O(n) part was that every time I call `initialize` I have to potentially loop over all the arrays declared over a domain before I confirm that `arr` is owned by the resized dom.

Reply:
Ah I'm stupid, I just realized you're saying we can remove the loop.